Lambda-encoding booleans (#2)
robrix/robrix.github.io: _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md

---
layout: post
title:  "All you need is λ, part one: booleans"
date:   2020-03-29 20:17:26
categories: ["lambda calculus", "plt"]
---

# All you need is λ, part one: booleans

> Nearly a hundred years ago, Alonzo Church invented computation in the form of the simple, elegant, and utterly inscrutable lambda calculus. Along with Alan Turing, he then proved that anything computable can be computed in the lambda calculus. However, nearly as soon as we had digital computers, we started inventing programming languages, and with them a vast treasure of features, beautiful and terrible, many of which seem very hard to relate to the fundamental nature of computability, let alone the lambda calculus.

Reviewer: Rather than

> Along with Alan Turing…

which could imply that they worked together on the lambda calculus, which isn't quite true, I would say

> Just as Alan Turing proved that anything computable can be computed with a Turing machine, so did Church with the lambda calculus…

or something along those lines. You could mention the Church-Turing thesis as well, if you wanted.

I would also drop the `utterly inscrutable`, since I don't think that's true! It's just unfamiliar to most people.

Author: The proof that the lambda calculus is Turing-complete—Church-Turing—is precisely what that sentence is describing.

I stand by my description of the lambda calculus as utterly inscrutable; in the absence of a metalanguage, it’s like reading binary machine code or something.

Reviewer: Does the Church-Turing thesis technically have a proof, since it's an informal statement? Anyway, I thought it would be nice to call it out by name, but your call.

And fair enough. It cannot be said to leap off the page, though I'd say it's easier than x86 assembly. 😆

Reviewer 2: maybe `elusive` instead of `utterly inscrutable`? I do think unfamiliarity with notation makes it challenging to decipher.

Reviewer 3: ```
Along with Alan Turing, he then proved that anything computable can be computed in the lambda calculus.
```

Fwiw I also think the section would benefit from a few rounds of refinement. I found the redundancy of the `compute` root in this sentence distracting.

While it’s true that anything which can be computed, period, can be computed in the lambda calculus, you might not want to: it’s austere, to say the least, and was not designed with modern sensibilities regarding readability in mind. We developed all those languages and features for a reason! Still, Church demonstrated not just that it was possible to compute anything computable with the lambda calculus, but also _how_ one might do so.

## Unconditional λ

> The lambda calculus’ austerity is extreme: you don’t even have booleans. All you have are:

Reviewer: ```suggestion
The lambda calculus’s austerity is extreme: you don’t even have booleans. All you have are:
```

Author: This is intentional.

Reviewer: Strictly speaking, it's incorrect; suffixed `'` plurals only apply if the word itself is plural _and_ ends in an `s`, e.g. it's `bus's`, `mice's`, but `cheerleaders'` (see _Dreyer's English_, p79). It's not the end of the world, though.

Reviewer 3: `calculi's`

Reviewer: That would work too.

> The lambda calculus’ austerity is extreme: you don’t even have booleans. All you have are:

Reviewer 2: This is a personal style preference for readability that would feel redundant to some, but I generally prefer numerated/bulleted ideas introduced before their explanations are introduced. This gives my brain a primer of 1) how many things I'm about to read about and 2) what their names are _before_ I learn about them. 

Something like:

> All you have are [Single-argument functions](), [Function application](), [Variables]()

followed by longer definitions/explanations or:

> All you have are:
> 1. [Single-argument functions]()
> 2. [Function application](URL)
> 3. [Variables](url)

and then their longer explanations.

1. Single-argument functions (“lambda abstractions,” but “lambdas,” “abstractions,” and “functions” will also be used interchangeably).

>     These are written `λ x . y`, for variable `x` and expression `y`, where `x` is now available as a bound variable in the body, and any enclosing definition of `x` is shadowed. (We shall assume strictly lexical scoping for the time being.)

Reviewer 2: > and any enclosing definition of `x` is shadowed

maybe provide a small example in parens?

>     In Haskell, we would write `\ x -> y` instead; in JavaScript, `function (x) { return y }` or `(x) => y`.

Reviewer 2: 👏🏼✨love examples that cross language ecosystem boundaries

2. Function application (“lambda application,” or “application” or “function calls”).

>     Applications are written as `x y`, for expressions x and y, and left-associated, i.e. `a b c` = `(a b) c` ≠ `a (b c)`. Function application binds tighter than lambda abstraction, i.e. `λ x . λ y . y x` = `λ x . λ y . (y x)` ≠ `λ x . (λ y . y) x`.

Reviewer: Maybe a table comparing LC/Haskell/JS notation, or perhaps a side-by-side implementation of const and id, would be appropriate? Your call.

Author: `const` and `id` don’t entail application, which is a shame. `fix` is probably too arcane. Hmm.

I was also thinking of maybe giving corresponding implementations for the `Bool` stuff in Haskell (& perhaps JS) but maybe that’s better left to the next post?

Reviewer: I think that table, placed near the end of the piece as an act of summarization, would be good.

>     The syntax is the same in Haskell; in JavaScript, we would write `x(y)` or `a(b, c)`. Note however that since lambda calculus functions are all single-argument functions, a more direct (tho less idiomatic) equivalent for the latter would be `a(b)(c)`.

Reviewer: ```suggestion
    The syntax is the same in Haskell; in JavaScript, we would write `x(y)` or `a(b, c)`. Note however that since lambda calculus functions are all single-argument functions, a more direct (though less idiomatic) equivalent for the latter would be `a(b)(c)`.
```

> 3. Variables (written as names; bound by enclosing lambdas; no `let` bindings, no globals; we will feel free to give variables names both great and small, and will include non-alphanumeric characters in names as we see fit, since the paucity of syntax means there’s little risk of ambiguity).

Reviewer: This parenthetical is so long that I'd say you should nix it and let the sentences within it flourish. You have a lovely prose style; let it flow.

> 3. Variables (written as names; bound by enclosing lambdas; no `let` bindings, no globals; we will feel free to give variables names both great and small, and will include non-alphanumeric characters in names as we see fit, since the paucity of syntax means there’s little risk of ambiguity).

Reviewer 2: ```suggestion
Variables. These are written as names; bound by enclosing lambdas; no `let` bindings, no globals; we will feel free to give variables names both great and small, and will include non-alphanumeric characters in names as we see fit since the paucity of syntax means there’s little risk of ambiguity.
```

> Since a lambda is the only way to introduce a value—it’s the only “literal” syntax, and there are no predefined global definitions or standard library—we can also see that the only kinds of runtime values must be closures, perhaps consisting of the name of the introduced variable, body of the lambda, & values of any variables it closed over when made (again, we assume strict lexical scoping).

Reviewer 2: This sentence is long and a little hard to read. Perhaps consider splitting it into two sentences:

```suggestion
Since a lambda is the only way to introduce a value—it’s the only “literal” syntax, and there are no predefined global definitions or standard library. We can also see that the only kinds of runtime values must be closures, perhaps consisting of the name of the introduced variable, body of the lambda, and values of any variables it closed over when made (again, we assume strict lexical scoping).
```

> How do you _do_ anything when you don’t even have `true` and `false`? Lambdas and variables don’t _do_, they merely _are_, so that leaves application. When all you have is lambda application, everything looks like a lambda abstraction, so we’ll represent booleans using lambdas.

Reviewer 2: This is an important paragraph since it prompts the reader to think and establishes a critical segue. However, it feels a little abrupt and could use a better transition to start it off and make it more harmonious.

> > Aside: Due to the lambda calculus’ terseness, I will be making free use of several notational conveniences:

Reviewer: I wouldn't necessarily foist this list into a blockquote, but I'll reserve judgment until I see it rendered on the page.

>
> 1. writing `λ x y . z` as an abbreviation of `λ x . λ y . z`.
> 2. writing `?` to stand for bits we don’t know yet, as tho we had an environment supporting holes.
> 3. writing definitions as tho we had a metalanguage.
> 4. referencing definitions elsewhere as tho we had globals.
> 5. writing type signatures as tho we had a type system, and even a typechecker, with as much polymorphism and inference as is convenient at any particular moment.
> 6. using syntactic recursion as tho it existed.
> 7. using general recursion as tho it made sense.

> > 8. ignoring application order, normalization, reduction, substitution, values, references, allocation, copying, space, time, entropy, and any and all other such details whenever I feel like it.

Reviewer: `s/tho/though` here (English orthography is a fuck but it's our unhappy burden to bear)

Author: I dunno, man, language evolves. We don’t write “although” as much as we used to, for example.

Reviewer: I think a full `though` fits the tone of this better; it's mostly-formal writing, after all. It somewhat interrupted the flow for me; language indeed evolves, but formal writing is not the best petri dish.

>

> > Further, I will by convention name types in `TitleCase` and both term and (local) type variables in `camelCase`.

Reviewer 2: ```suggestion
> Further, I will abide by convention name types in `TitleCase` and both term and (local) type variables in `camelCase`.
```

Author: “name” is a verb here; “I will, by convention, name types in …”

>

> > I will try to avoid pulling rabbits from hats too wantonly, but for now, I’ll ask you to suspend disbelief; I hope to revisit and justify some of these in later posts.

Reviewer: I don't think you necessarily need this disclaimer. The above syntactic sugar is hardly excessive.

Author: I mostly want to indicate that I intend to justify even more reduction of syntax into the lambda calculus.

> ## λ is blind

Reviewer 2: I love these and the title of the post so much.

```haskell
if_ :: Bool -> a -> a -> a
if_ cond  then_ else_ = if cond then then_ else else_
```

In a strict language like JavaScript, we’d instead take functions for the alternatives:

```javascript
function if_(cond, then_, else_) {
  if (cond) {
    then_();
  } else {
    else_();
  }
}
```

Both these definitions use the language’s native booleans and `if` syntax (a tactic for implementing embedded DSLs known as “meta-circularity”), and thus aren’t viable in the lambda calculus. However, they do give us a hint: in both cases we have a function taking a condition, consequence, and alternative, and using the first to select one of the latter two. In the lambda calculus, we might start by writing:

```
if = λ cond then else . ?
```

(Note: there aren’t any keywords in the lambda calculus, so there’s nothing stopping me from naming variables things like `if`, a fact which I will take free advantage of.)

> We’ve introduced a definition for `if`, as a function of three parameters; now what do we do with them? The lambda calculus’ stark palette makes it easy to enumerate _all_ the things we can do with a variable:

Reviewer: ```suggestion
We’ve introduced a definition for `if`, as a function of three parameters; now what do we do with them? The lambda calculus’s stark palette makes it easy to enumerate _all_ the things we can do with a variable:
```

> 1. Ignore it, or shadow it with another lambda binding the same name.

Reviewer: ```suggestion
1. Ignore it, or shadow it with another lambda binding of the same name.
```

> 1. Ignore it, or shadow it with another lambda binding the same name.

Author: Binding is used as a verb here, not a noun.

Reviewer: Fair!

> 1. Ignore it, or shadow it with another lambda binding the same name.

Reviewer 2: Example?

> 2. Mention it (perhaps nested), whether on its own in the body of a lambda, or somewhere on either side of an application (or both).

Reviewer: Can you expand on precisely what you mean by `(perhaps nested)`?

> 2. Mention it (perhaps nested), whether on its own in the body of a lambda, or somewhere on either side of an application (or both).

Reviewer 2: Example?

We could for example simply return `then` or `else`:

```
if = λ cond then else . then
if = λ cond then else . else
```

But in that case the conditional isn’t conditional at all—the value in no way depends on `cond`. Clearly the body must make use of all three variables if we want it to behave like the `if`s we know and love from other languages.

Taking a step back for a moment, let’s examine the roles of `if`’s arguments. `then` and `else` are passive; we only want to use or evaluate one or the other depending on the value of `cond`. `cond`, then, is the key: it takes the active role.

Thus, in the same way that our `if_` functions in Haskell & JavaScript employed those language’s features to implement, we’re going to define `if cond then else` as the application of the condition to the other two parameters:

```
if = λ cond then else . cond then else
```

> > Aside: under standard semantics for the lambda calculus, we could simplify this definition further by “η-reducing” it, noting that `λ x . f x` behaves the same as `f` alone. Nevertheless, we will leave this and likely other definitions in their full, η-long forms for the sake of clarity.

Reviewer 2: Maybe link to a reference defining  eta-reduction?

> This feels strangely like cheating: surely we’ve only moved the problem around. Now instead of `if` making the decision about which argument to return, we’ve deferred it to `cond`. But `if` and `cond` aren’t the same kind of thing, semantically; `if` takes a boolean and two other arguments and returns one of the latter, while `cond` _is_ a boolean—albeit evidently a boolean represented as a function. Let’s make that precise by writing down `if`’s type:

Reviewer: s/kind of thing//

```
Bool = ∀ a . a -> a -> a
```

> Aside: I’m making explicit use of the for-all quantifier here to drive home the point that any particular `Bool` value must be able to be applied to `then` and `else` values of any arbitrary type `a`, defined now or in the future.
>
> By the same token, we could have written `if`’s type more explicitly as:
>
> ```
> if : ∀ a . Bool -> a -> a -> a
> ```
>
> Here and in future, local type variables can be assumed to be implicitly generalized in the same manner as Haskell if not otherwise quantified.

If a given `Bool` is a function of two arguments of arbitrary type, returning the same type, it must therefore select one of its arguments to return. There are only two distinguishable inhabitants of `Bool`, `true` and `false`, so we can therefore deduce that since `if` defers the selection of the result to the `Bool`, for `true` and `false` to actually differ they must make opposite selections. In other words, `true` must return the `then` parameter, while `false` must return the `else` one:

```
true, false : Bool
true  = λ then else . then
false = λ then else . else
```

> We didn’t move the problem around at all—we actually solved it (once we defined `true` and `false`, anyway). It’s simply that encoding booleans in this manner makes `if` largely pointless, since if we can apply `if` to a `Bool` and two arguments, we could equally apply the `Bool` to those arguments directly.

Reviewer: ```suggestion
We didn’t move the problem around at all—we solved it (once we defined `true` and `false`, anyway). The fact is that encoding booleans in this manner makes `if` pointless, since if we can apply `if` to a `Bool` and two arguments, we could equally apply the `Bool` to those arguments directly.
```

> > Aside: we chose to define `if` as applying the `Bool` to the other arguments in the same order it received them, but we could just as easily have swapped them:

Reviewer: If you have footnote support in your blog, these asides would be excellent as footnotes (or even better, sidenotes a là Edward Tufte, but that's a bit fiddly to get working).

Author: I’m actually pretty 👎 on footnotes, but I was thinking of literally pulling them into an adjacent column, or collapsing them using `<detail>`.

Reviewer: Adjacent column would be fine. It's what I do on my blarg.

>
> ```
> if = λ cond then else . cond else then
> ```
>
> In this case, `if` would be more useful since it would preserve our familiar argument ordering. As an exercise for the reader, consider what other effects this difference would have. What are the tradeoffs, syntactically and semantically? When would one or the other definition be more or less convenient?

It’s frequently convenient to conflate booleans with bits, their minimal representation, but in truth they’re not the same at all. Practically, some programming languages define booleans as a byte in memory, perhaps clamping its values to 0 and 1; others define them as instances of some boolean class, or constructors of an algebraic datatype. Some provide no formal relationship between `true` and `false` at all, save for a common interface—duck typing.

Mathematically, booleans are the values in propositional logic; the upper and lower bounds of a lattice; the zero and one of a semiring; the members of the set with cardinality 2; and many other things in many different contexts.

> 

Reviewer: I would emphasize `representing it with functions which can perform all the operations possible with the original datatype`, since it's the linchpin of this piece.

Operationally, booleans represent choice, and this is a pattern that we’ll see repeated: encoding a datatype with lambdas means representing it with functions which can perform all the operations possible with the original datatype. All operations on booleans can be defined by selecting between two alternatives, which is precisely what our encoding does.

We can demonstrate this by defining some other operations on booleans, e.g. logical operators, using the encoding we’ve built thus far.

`not` takes a single `Bool` and returns another:

```
not : Bool -> Bool
not = λ x . ?
```

As when defining `if`, all we can do with a `Bool` is branch on it:

```
not = λ x . if x ? ?
```

> > Aside: as discussed in a previous aside, `if` is operationally redundant—i.e. `if x y z` is operationally equivalent to `x y z`—given the ordering of arguments to `Bool`s which we selected earlier. It is, however, pleasantly evocative, and so is used for clarity and so we can stop talking about that ordering decision.

Reviewer: I never realized this. So cool!

```
or, and : Bool -> Bool -> Bool
or  = λ x y . ?
and = λ x y . ?
```

As with `not`, all we can do with `Bool`s is branch:

```
or  = λ x y . if x ? ?
and = λ x y . if x ? ?
```

For `or`, if `x` is `true`, we can return `true` immediately (“short-circuiting”). For `and`, it’s the opposite:

```
or  = λ x y . if x true ?
and = λ x y . if x ?    false
```

If `x` is `false`, `or` needs to test whether `y` is `true`; likewise, if `x` is `true`, `and` needs to test whether `y` is also `true`. Once more, all we can do with `Bool`s is branch:

```
or  = λ x y . if x true       (if y ? ?)
and = λ x y . if x (if y ? ?) false
```

And since we must return a `Bool`, we can use `true` and `false`:

```
or  = λ x y . if x true              (if y true false)
and = λ x y . if x (if y true false) false
```

Pleasantly, `if y true false` (and likewise `y true false`) is operationally equivalent to `y`. Using that equivalence, we can simplify these definitions, leaving us with:

```
or  = λ x y . if x true y
and = λ x y . if x y    false
```

> Aside: as an exercise, define `xor : Bool -> Bool -> Bool`.

## Conclusion

> In this post, we’ve explored defining a ubiquitous programming language feature—booleans—using nothing more than the spartan trappings of the lambda calculus, and emerged with a language which can express not merely functions and their applications, but also fundamental metaphysical concepts such as truth.

Reviewer 2: I'm a fan of shorter sentences.

```suggestion
In this post, we’ve explored defining a ubiquitous programming language feature—booleans—using nothing more than the spartan trappings of the lambda calculus. We have emerged with a language which can express not merely functions and their applications, but also fundamental metaphysical concepts such as truth.
```
